### DISCUSSION TITLE
Zen Cart Roadmap

### DISCUSSION THREAD

**Discussion by zcwilt** [zencart/zencart#5090] [General]
Currently we have no roadmap for V158 and beyond apart from ongoing bug fixing and PHP compatibility.

What should be in the roadmap ? 

Some suggestions would be

 - adding some plugins to core 
 - developing new features 
 - refreshing the template for both catalog and admin 
 - updating core code to use more modern features
 -  Maybe converting more plugins to the new encapsulated standard.

Others might have different ideas.

If you want to suggest a plugin to add to core, make sure you point to the plugin URI. I'll try an keep up with this to update a list, and maybe add a poll for this.

If you want to suggest other core changes etc, it would be better if you created a new discussion and linked that discussion here, so that we don't get to distracted discussing them in this thread.

I'd rather not see posts just saying lets include xxxx.plugin, but rather including a business reason why that plugin would benefit being in core.

**Comment by njcyx** [comment]
First of all, PHP7.4 goes EOL on Nov 28, 2022. Now it is already August, so there is only 3 more months left. All the ZC users are waiting anxiously for the final release of 1.5.8 for PHP 8.0.

Then PayPal module upgrade. Lots of European ZC users are requesting the support for 3D SECURE compatible. Now it looks like only @lat9 is working on this module with her limited resources. European countries enforced 3D SECURE already but US has not enforced it yet. I'm one of the end user who mainly uses PayPal. I'm not sure if zen cart will drop the paypal support in the future. Instead, Square is recommended here and there. Deleted (It seems that there is an ongoing, urgently and important update for square module as well.). 

We are so close to the "due day".

**Reply by proseLA** [reply]
> Instead, Square is recommended here and there. It seems that there is an ongoing, urgently and important update for square module as well.


as the developer who coded the new square_webPay module for zen cart, i am not aware of any ongoing or important updates that are necessary.  perhaps you can be specific as to what is so urgently needed.

**Reply by mc12345678** [reply]
> First of all, PHP7.4 goes EOL on Nov 28, 2022. Now it is already August, so there is only 3 more months left. All the ZC users are waiting anxiously for the final release of 1.5.8 for PHP 8.0.

The latest subversion of Zen Cart 1.5.7 works with php 8.0. As always, Zen Cart users anxiously await the next release. This portion is a non-issue. The part that follows, that is certainly important.

**Reply by njcyx** [reply]
@mc12345678 As stated here https://docs.zen-cart.com/user/first_steps/server_requirements/

Zen Cart v1.5.7 is designed for PHP 5.6 through PHP 8.0
(If you are using PHP 8.0, be sure to [suppress logging duplicate-language definitions](https://docs.zen-cart.com/user/troubleshooting/constant_already_defined/)).

I don't think "suppress logging duplicate-language definitions" is normal for a fully compatible product. At least for other early versions, users don't need to do this change. In my point of view, php 8.0 with zc 1.5.7 can only be regarded as “conditional compatible”.

Thanks for your tip though.

**Reply by njcyx** [reply]
> > Instead, Square is recommended here and there. It seems that there is an ongoing, urgently and important update for square module as well.
> 
> as the developer who coded the new square_webPay module for zen cart, i am not aware of any ongoing or important updates that are necessary. perhaps you can be specific as to what is so urgently needed.

@proseLA Maybe I'm wrong. I have edited the original post.

**Comment by mc12345678** [comment]
At one point the discussion was that catalog side plugin support would be incorporated or supported in 1.5.8, yet that does not appear to be operational at this time. What is the vision of plugins? The ability to retain multiple versions of the software to be installed and removed at will seems to have been desired.  What is the consideration of plugin download direct to store? A lot of code and features were brought into the plugin model, what was the inspiration? What is/was the skeleton concept?

**Comment by mc12345678** [comment]
Based on remarks and code edits made, it seems that existing notifiers need a thorough scrub to ensure that the variables associated with the intended usage of that notifier or those that become available only at that point are made available to an observer without use of (most) global variables.

**Reply by torvista** [reply]
You mean a notifier should pass every variable that may be required?
Why duplicate the global variables?

**Comment by mc12345678** [comment]
Moving zencart code out of templates and into variables to be presented by the template(s). E.g. path(s) forwards to encourage/obtain personal templates that can actually support the store and are supported by the community.

**Comment by torvista** [comment]
I'm taking this roadmap discussion as purely code development.
And I would point out that posters should keep an eye on the original post, to try and keep it on topic. The idea of a discussion is to arrive at a conclusion and instigate action quickly efficiently.

But, features and priority of inclusion should be a forum discussion, not hidden here. Discussions here are off the radar for most users.
The marketing plan must prioritize maintaining current users. The move to ZC158 requires a rebuild so the more popular plugins that are included, the less upgrade pain involved and the less chance of users looking elsewhere. 
Yes there needs to be a list (use comments here + the most downloaded plugins as a start) and a poll.

There needs to be large roadmap so it allows folks of varying ability/confidence to take ownership of an issue and state "I'll do that one".

## Plugins to Core   
1) The primary code functionality hole is no stock control of variants. This is covered by two plugins 

https://github.com/mc12345678/Stock_By_Attributes_Combined
https://vinosdefrutastropicales.com/index.php?main_page=product_info&cPath=2_7&products_id=46

but should be a core feature. How that may be resolved is another discussion.

2) Image handling. Uploading three images in 2022: please no!
Since there is only one plugin, that should be core now.

https://github.com/lat9/zen_Image-Handler

3) Structured data
Still a moving target? Not really. Few (no) changes have been required in months to meet changing Google requirements.

https://github.com/torvista/Zen_Cart-Structured_Data

## New Features
What do the other ecommerce solutions have that Zen Cart does not?

## Template
### Admin
Making it pretty: zero priority.
100% valid html and accessibility: yes. Including reorganisation of menus. Low priority and a big potential time-waster discussing what should go where. But necessary. 
My opinion on menus is that the core ones should be fixed (alphasorted) and plugins added below.

### Shopfront
Bin classic template.
Responsive classic: despite it being intended as an example only, it is the primary decider on whether a new user experiments further, so either this or another more flashy one does have importance if only from a marketing point of view.

## Updating core code to be more modern
Not really up for discussion. Of course. Soon as possible.  
But guidance required from the top as to what should/can be done piecemeal when a file is touched for another reason.  
Unfortunately php upgrades are continually forcing code fixes, Zen Cart needs to get ahead of that curve. It seems the only way is to use php 8.2beta for development!

## Converting plugins
If it's generally useful: add to core. It is more time-wasting to bugger about making a plugin work with no core edits than just PR it.
If it's a minority feature: down to plugin author to update/customer to pay for the work to update it.

ALL tasks needs to go on a roadmap to they can be picked off by whoever wants to do it. The more basic things that are tasks, the more that people may get involved.

**Reply by gunghonights** [reply]
Hi, I have read with interest the discussions in this thread and others about the road map and developments and future of zencart.   I have been a long time user of Zencart running a small store and have tried to do any changes and development myself to keep costs down, but am not a programmer.  I have struggled to upgrade and keep up with the developments and getting my stores plugins to work with newer versions has at times been fat too tricky.  I have thought many times of jumping ship.

I thought I would jump in here as one of the discussions has been about attracting users, more developers and getting younger fresher people to join.  I think I started reading the thread with the heading Zencart is Dead. 

I have elected to enter the discussion based on the fact that version 2 has a thread on the main site, as part of the roadmap developments.  I have just installed locally to take a look and there are many things that torvista  has mentioned above that I am not sure have been taken on board which I think will mean that Zencart will struggle to attract new users.  

1)  Stock by attributes - should now be a core feature - each time there has been an upgrade I have struggled to get this reinstalled again.
2) Image handing again should be core as it has been around as a plug in for so long now.
3) My second battle with Zencart is getting a design for my brand that reflects modern design and has many of the features that are available in the Westminster New Template.  I want to be able to move elements quickly but really struggle to work out which blocks I can move around.  Currently battling with Bootstrap in 1.5.8a but again features /plug-ins often clash and I have struggled/still struggling to do the upgrade.  again I agree with torvista.  we should have a responsive template for the modern age that allows plugins front end features to be dropped in.  I have found something as simple as swapping the menus, logo and login /basket blocks around challenging and as a store owner not a developer I want to be able to find/adjust templates and designs that work, not trying to battle with the code.
4) Super orders with printing of multiple orders /invoices at once and editing of orders by admin - I used to use up to 1.3.9 (I think) and then I could not get it to work and lost what was great efficient functionality.

I only have to look at wordpress and can find various templates that are contemporary designs at just seem to drop in and work.

Would love to be able to get more involved but do not have the coding skills and this may be part of the reason why if I have made some statements above that seem ignorant. - I can only apologise..... but there will be many other aspiring young business people who are great at business and marketing and not got the time to learn to code - they will feel like me, frustrated that I cannot achieve something quickly  and will want to move on to something that is allowing them to be successful.  

It will be ashame to have to change, as I too have invested many hours.  It would be a shame to see it fade and die.  Do let me know how I can help, contribute or better still learn how to code at the level needed to move the project forward.  I fear however I am a person with not enough time and perhaps not enough brains.

I may be reaffirming what you already know but I hope it helps.  My best regards to all the developers who have kept the project going and have helped my through the forum.  I look forward to the continued development and would love to help further.

**Reply by drbyte** [reply]
@gunghonights Thanks for your input. It is helpful to hear your perspective. Much appreciated!

The points you raise are some of the most complex things to maintain. But that doesn't make them any less important.

Again, thanks for the feedback: it does help us prioritize things!

**Reply by torvista** [reply]
>there are many things that torvista has mentioned above that I am not sure have been taken on board which I think will mean that Zencart will struggle to attract new users.

Sorry for the reality check but as far as **I** am aware (being part of the general public), absolutely nothing has changed apart from the usual piecemeal/semi-random code fettling and minor bug fixing.

V200 is about to be released, a major version change but which in reality contains zero new features or functionality.

This is absolutely not a **complaint** that "someone should do something", as though lazy developers are on the beach with their pina coladas instead of coding, although that's my plan later today.
Obviously everyone and anyone who "does things" is already doing what they can within their personal limits and in line with what they perceive as useful/easily done and my comments are no doubt extremely irritating: no-one likes being told what they already well-aware of.
But, the fact remains that until there is a real Roadmap of clearly-defined objectives and tasks of varying difficulty where people can take ownership of those tasks as per their abilities/interest/time, progress will remain glacial.

If you don't know your destination, how can you get there?

**Reply by KitchenWarehouse** [reply]
Long time user of ZenCart from first our first store in 2011 all the way to today, I agree most points and plugins above and mainly agree with Torvistas thoughts on a more clearly defined tasks and roadmap for further development.

In my personal thoughts on certain needs

**Plugins to Core**

1. Image Handler - Seems like one of the first plugins everyone would want to install when making a new store
2. Structured Data
3. Email order confirmation PDF
4. More advanced control over shipping - I have trouble configuring shipping via location and amount without using advanced shipper which as a paid plugin doesnt seem to be in active development. Having an overhaul of the current shipping modules into more useful ones would be amazing.

Im in no way a developer and have basic knowledge of php so I am reliant on the amazing developers to keep bring further updates to help run the store as fast an efficiently as possible. I would love to learn more to play a more active role in development.

**.zip plugins**

I recently installed the ModList plugin and it wasa drop-in install into the zc_plugins folder which was amazing, I had never seen this before and when looking into it it seems to be admin only currently, I would love to see a push to move to this standard for both admin and catalog side plugins if feasible as I believe it would make maintainence easier.

**Github**
This is a small one but I would like to see either a more prominent link to the github on the forum which is where quite a bit of the development discussion goes on

**Roadmap**
With the above thoughts on the roadmap and the need for a clearly defined set of objects that we agree should be worked on what would be the best suggestion to get this implemented? Could we use the issue tracker of github?

Could we have a forum post asking users what they would like to see in future versions of zencart?

Thanks for all the work you guys do!

**Reply by gunghonights** [reply]
@KitchenWarehouse - I agree with your point 4 - the  Advanced Shipping by CEON (which I think is the one you refer to) was great but paid for but its features should be core functionality.
 In addition to this one of the features I wanted at the start when I originally set up my business was to integrate with either a proper stock management system or an accounts package.  I paid for service for a while but then it got too expensive every time there was an update, either with zencart or the accounting package.   Not sure the answer on that bit specifically unless we link up with another open source accounts package and at least offer the export of data in a certain structure to help that integration - this may be what is alluded to with structured data point 2 above?  Do clarify, please.  
The data export plugin (which I have not tried yet, was going to be my next attempt and to then use MS Access or other database of some kind to do the stock management but time and knowledge has limited me here.   A recommended solution here might help persuade users to choose zencart if that part of the process works smoothly.

Hope this helps with either more ideas to get the direction and roadmap going or at least to start a discussion, even if ultimately it gets rejected.

**Reply by torvista** [reply]
Any discussion is a good thing, but it always goes this way...in all directions then peters out....without any focus on where Zen Cart wants to be/should be in the marketplace and how best to get there.
Getting into discussions about plugins etc. is looking at details instead of the big picture.

I don't think I can improve on my previous comment here: https://github.com/zencart/zencart/discussions/4805#discussioncomment-2806950

**Reply by gunghonights** [reply]
@torvista   just read https://github.com/zencart/zencart/discussions/4805#discussioncomment-2806950 and yes you sum it up exactly.  Did the spreadsheet of features and comparisons ever get generated?  it is in a discussion I have not found yet?

**Reply by torvista** [reply]
Someone made an effort: https://github.com/zencart/zencart/discussions/4805#discussioncomment-2828595

I think the whole thread contains multiple positive suggestions and is worth reading.
The ideas are not in short supply, just the man/woman-power....hence the need for focus and inviting help on small, concrete tasks.

**Comment by zcwilt** [comment]
@torvista thanks for the update, I'll try and pull across some stuff from https://github.com/zencart/zencart/discussions/4805
in terms of other peoples ideas about what should be in core.

Building on that we then need to roadmap stuff to ensure we are not just bugfixing.

**Reply by torvista** [reply]
In the interest of advancing this stalled topic with minimum effort, I noticed that Prestashop apparently use the GitHub Issues->Features to add feature request and allow upvotes and so create a sort of roadmap, but I fail to see how to view/list Issues tagged as Features in Zen Cart or Prestashop to be able to vote on them.
